ESQL: only 400 response on IllegalArgumentException in the compute engine

To reproduce: Check out main before https://github.com/elastic/elasticsearch/pull/125764 was merged. Run a local ES instance via `./gradlew run` and ingest the csv test data:
```
./gradlew :x-pack:plugin:esql:qa:testFixtures:loadCsvSpecData --args="http://elastic-admin:elastic-password@localhost:9200"
```
Run the following query (until it errors, it's slightly non-deterministic):
```
from *    
| enrich languages_policy on author.keyword
| sort book_no
| limit 1
```
This gives a 400 when it errors. Running with `error_trace` shows that it's due to a mismatch in data types (originally from a wrong layout):
```
 Expected [BYTES_REF] but was [INT]
...
Caused by: java.lang.IllegalArgumentException: Expected [BYTES_REF] but was [INT]
  \tat org.elasticsearch.compute.operator.topn.ValueExtractor.extractorFor(ValueExtractor.java:30)
  \tat org.elasticsearch.compute.operator.topn.TopNOperator$RowFiller.<init>(TopNOperator.java:150)
  \tat org.elasticsearch.compute.operator.topn.TopNOperator.addInput(TopNOperator.java:371)
  \tat org.elasticsearch.compute.operator.Driver.runSingleLoopIteration(Driver.java:276)
...
```
I believe this IAE should've triggered a 500 response; returning 400 can obscure a bug in the compute engine.

--- a/docs/changelog/126296.yaml
+++ b/docs/changelog/126296.yaml
@@ -1,0 +1,5 @@
+pr: 126296
+summary: Fail with 500 not 400 for `ValueExtractor` bugs
+area: ES|QL
+type: bug
+issues: []

--- a/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/ValueExtractor.java
+++ b/x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/topn/ValueExtractor.java
@@ -27,7 +27,10 @@ interface ValueExtractor {
 
     static ValueExtractor extractorFor(ElementType elementType, TopNEncoder encoder, boolean inKey, Block block) {
         if (false == (elementType == block.elementType() || ElementType.NULL == block.elementType())) {
-            throw new IllegalArgumentException("Expected [" + elementType + "] but was [" + block.elementType() + "]");
+            // While this maybe should be an IllegalArgumentException, it's important to throw an exception that causes a 500 response.
+            // If we reach here, that's a bug. Arguably, the operators are in an illegal state because the layout doesn't match the
+            // actual pages.
+            throw new IllegalStateException("Expected [" + elementType + "] but was [" + block.elementType() + "]");
         }
         return switch (block.elementType()) {
             case BOOLEAN -> ValueExtractorForBoolean.extractorFor(encoder, inKey, (BooleanBlock) block);

